ledger/participant-integration-api/src/main/scala/platform/store/backend/h2/H2StorageBackend.scala

```diff
@@ -238,4 +252,133 @@ private[backend] object H2StorageBackend
       pruneAllDivulgedContracts: Boolean,
       connection: Connection,
   ): Unit = ()
+
+  override def maximumLedgerTime(
+      ids: Set[ContractId]
+  )(connection: Connection): Try[Option[Instant]] = {
+    if (ids.isEmpty) {
+      Failure(emptyContractIds)
+    } else {
+      def lookup(id: ContractId): Option[Option[Instant]] = {
+        import com.daml.platform.store.Conversions.ContractIdToStatement
+        SQL"""
+  WITH archival_event AS (
+         SELECT 1
+           FROM participant_events_consuming_exercise, parameters
+          WHERE contract_id = $id
+            AND event_sequential_id <= parameters.ledger_end_sequential_id
+          FETCH NEXT 1 ROW ONLY
+       ),
+       create_event AS (
+         SELECT ledger_effective_time
+           FROM participant_events_create, parameters
+          WHERE contract_id = $id
+            AND event_sequential_id <= parameters.ledger_end_sequential_id
+          FETCH NEXT 1 ROW ONLY -- limit here to guide planner wrt expected number of results
+       ),
+       divulged_contract AS (
+         SELECT null
+           FROM participant_events_divulgence, parameters
+          WHERE contract_id = $id
+            AND event_sequential_id <= parameters.ledger_end_sequential_id
+          ORDER BY event_sequential_id
+            -- prudent engineering: make results more stable by preferring earlier divulgence events
+            -- Results might still change due to pruning.
+          FETCH NEXT 1 ROW ONLY
+       ),
+       create_and_divulged_contracts AS (
+         (SELECT * FROM create_event)   -- prefer create over divulgence events
+         UNION ALL
+         (SELECT * FROM divulged_contract)
+       )
+  SELECT ledger_effective_time
+    FROM create_and_divulged_contracts
+   WHERE NOT EXISTS (SELECT 1 FROM archival_event)
+   FETCH NEXT 1 ROW ONLY"""
+          .as(instantFromMicros("ledger_effective_time").?.singleOpt)(connection)
+      }
+
+      val queriedIds: List[(ContractId, Option[Option[Instant]])] = ids.toList
+        .map(id => id -> lookup(id))
+      val foundLedgerEffectiveTimes: List[Option[Instant]] = queriedIds
+        .collect { case (_, Some(found)) =>
+          found
+        }
+      if (foundLedgerEffectiveTimes.size != ids.size) {
+        val missingIds = queriedIds.collect { case (missingId, None) =>
+          missingId
+        }
+        Failure(notFound(missingIds.toSet))
+      } else Success(foundLedgerEffectiveTimes.max)
+    }
+  }
+
+  override protected def activeContract[T](
```

Comment:
this is a creative approach to override indeed, but I would like to convince you to take another because:
* not exactly DRY, quite some copy paste takes place
* in case of activeContract it overrides an intricate part of the contract implementation. I think this is quite dangereous
* you can think of inheritance here as a side effect, not the goal, it is already on the table to switch to composition in StorageBackend

therefore I would suggest to organize the ContractStorageBackend, that it has template methods defining the sql literals, and override those only in H2, and not overriding maximumLedgerTime and especially activeContract (also keeping private stuff private of course)

happy to elaborate on a call

Reply:
I agree, this can definitely be improved. This case is another argument for removing the `participant_events` view at all.

Reply:
I've changed the approach and have modified queries on the common level instead of modifying them specifically for H2. This removes the issues you mentioned.